### PR TITLE
Fix #1374: Personalized FIDO2 assertion challenge

### LINKED ISSUE #1374: Personalized FIDO2 assertion challenge
Currently, we allow fetching an anonymous assertion challenge. We should allow personalizing the challenge immediately. As a result, we can return list of supported FIDO2 authenticators right away and ensure future check of the authenticator that was actually used against allowed authenticators.

### PR DESCRIPTION
**Pull request by petrdvorak**

### CODE REVIEW

**Review by romanstrobl — APPROVED** [review]
Looks OK.

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/AssertionChallengeConverter.java [hunk 1]
```diff
@@ -48,6 +55,65 @@ public AssertionChallengeResponse fromChallenge(AssertionChallenge source) {
...
+        destination.setApplications(source.getApplicationIds());
+        destination.setTemplateName(source.getTemplateName());
+        destination.getParameters().putAll(source.getParameters());
+
+        if (authenticatorDetails != null && !authenticatorDetails.isEmpty()) {
+            final Set<String> allowCredentials = new LinkedHashSet<>();
+            for (AuthenticatorDetail ad : authenticatorDetails) {
+                allowCredentials.add(ad.getExternalId());
+            }
+            destination.setAdditionalData(Map.of("allowCredentials", allowCredentials));
```
Comment (zcgandcomp) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/AssertionChallengeConverter.java:82]: If possible do not put this logic - really on the `activationId`. But at least put proper externalized constant, i.e., ATTR_ALLOW_CREDENTIALS -  with modification of visibility.

**Review by zcgandcomp — CHANGES_REQUESTED** [review]

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/service/AssertionService.java [hunk 2]
```diff
@@ -75,12 +77,25 @@ public AssertionService(CryptographyService cryptographyService, AuthenticatorPr
...
     public AssertionChallengeResponse requestAssertionChallenge(AssertionChallengeRequest request) throws Exception {
-        final AssertionChallenge assertionChallenge = assertionProvider.provideChallengeForAssertion(
-                request.getApplicationIds(), request.getTemplateName(), request.getParameters(), request.getExternalId()
-        );
+
+        final List<AuthenticatorDetail> authenticatorDetails = new ArrayList<>();
+
+        // If user ID is specified, fetch the user authenticators that should be allowed to respond the challenge
+        final String userId = request.getUserId();
+        if (userId != null) {
```
Comment (zcgandcomp) [inline powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/service/AssertionService.java:85]: We should move this logic to the provider. 
As of now:

- This service gets a list of possible authenticators from the provider. 
- Sends the list to the provider.
- The provider adds the list to his response.

Getting the list of possible authenticators and adding them to the response on the same level will increase readability.

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/AssertionChallengeConverter.java (hunk 1 above)
Comment (petrdvorak) [reply powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/rest/model/converter/AssertionChallengeConverter.java:82]: The change to activation IDs needs to be done later, it seems too fragile to fix in the same change.

File: powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/service/AssertionService.java (hunk 2 above)
Comment (petrdvorak) [reply powerauth-fido2/src/main/java/com/wultra/powerauth/fido2/service/AssertionService.java]: Didn't improve readability much, but OK

**Review by romanstrobl — DISMISSED** [review]
Nice improvements!

**Review by zcgandcomp — DISMISSED** [review]
Acceptable.

**Review by romanstrobl — APPROVED** [review]
Approved.

**Pull request merged**